app/controllers/multifactor_auths_controller.rb

```diff
@@ -21,14 +27,55 @@ def create
     else
       flash.now[:success] = t(".success")
       @continue_path = session.fetch("mfa_redirect_uri", edit_settings_path)
-      session.delete("mfa_redirect_uri")
-      render :recovery
+
+      if current_user.mfa_device_count_one?
+        session[:show_recovery_codes] = true
+        redirect_to recovery_multifactor_auth_path
+      else
+        redirect_to @continue_path
+        session.delete("mfa_redirect_uri")
+      end
     end
   end
 
   def update
+    session[:level] = level_param
+    @user = current_user
+
+    setup_mfa_authentication
+    setup_webauthn_authentication
+
+    create_new_mfa_expiry
+
+    render template: "multifactor_auths/mfa_prompt"
+  end
+
+  def mfa_update
+    if current_user.ui_mfa_verified?(params[:otp])
+      update_level_and_redirect
+    else
+      redirect_to edit_settings_path, flash: { error: t("multifactor_auths.incorrect_otp") }
+    end
+  end
+
+  def webauthn_update
+    @user = current_user
+    unless @user.webauthn_enabled?
+      redirect_to edit_settings_path, flash: { error: t("multifactor_auths.require_webauthn_enabled") }
+      return
+    end
```

Comment:
Not necessary, but this could be replaced with something like:
```ruby
before_action :require_webauthn_enabled, only: webauthn_update
```

```ruby
def require_webauthn_enabled
    return if current_user.webauthn_enabled?

    delete_mfa_level_update_session_variables
    redirect_to edit_settings_path, flash: { error: t("multifactor_auths.require_webauthn_enabled") }
  end
```